source/system_info.c

```diff
@@ -0,0 +1,69 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#include <aws/common/private/system_info_priv.h>
+
+#include <aws/common/logging.h>
+
+struct aws_system_environment *aws_system_environment_load(struct aws_allocator *allocator) {
+    struct aws_system_environment *env = aws_mem_calloc(allocator, 1, sizeof(struct aws_system_environment));
+    env->allocator = allocator;
+
+    if (aws_system_environment_load_platform_impl(env)) {
+        AWS_LOGF_ERROR(
+            AWS_LS_COMMON_GENERAL,
+            "id=%p: failed to load system environment with error %s.",
+            (void *)env,
+            aws_error_debug_str(aws_last_error()));
+        goto error;
+    }
+
+    AWS_LOGF_TRACE(
+        AWS_LS_COMMON_GENERAL,
+        "id=%p: virtualization vendor detected as \"" PRInSTR "\"",
+        (void *)env,
+        AWS_BYTE_CURSOR_PRI(aws_system_environment_get_virtualization_vendor(env)));
+    AWS_LOGF_TRACE(
+        AWS_LS_COMMON_GENERAL,
+        "id=%p: virtualization product name detected as \"" PRInSTR " \"",
+        (void *)env,
+        AWS_BYTE_CURSOR_PRI(aws_system_environment_get_virtualization_vendor(env)));
+
+    env->os = aws_get_platform_build_os();
+    env->cpu_count = aws_system_info_processor_count();
+    env->cpu_group_count = aws_get_cpu_group_count();
+
+    return env;
+error:
+    aws_mem_release(allocator, env);
```

Comment:
trivial: I like the pattern of just calling the destroy() function, so that if you're halfway through setup when it fails, you still tear down every individual member
```suggestion
    aws_system_environment_destroy(env);
```